File: scripts/find-closest-words.js
import fs from "fs"
import { makeDirectory } from "wordpaths-common/src/Common_.js" 
import Encoder_ from "wordpaths-common/src/Encoder_.js"
import Array_ from "wordpaths-common/src/Array_.js"
import Annoy from "annoy"

function normalize(vector) {
    const magnitude = Math.sqrt(vector.reduce((sum, val) => sum + val * val, 0));
    return vector.map(val => val / magnitude);
}

console.log("CLOSEST WORDS FINDER")
console.log("========================================================")

const MODEL_FOLDER    = "./data/glove-50d/preprocessed"
const OUTPUT_FOLDER   = "./data/glove-50d/preprocessed"

const VOCABULARY_FILE       = MODEL_FOLDER + "/vocabulary.txt"
const VECTORS_FILE          = MODEL_FOLDER + "/vectors.bin"
const CEFR_DATA_FILE        = "./data/cefr/cefr-data.json"
const STOP_WORDS_FILE       = "./data/stopwords/stopwords.txt"

// --- load vectors 
console.log("Loading vectors.")
let vectorBytes      = fs.readFileSync(VECTORS_FILE)
let vectors          = Encoder_.decodeFloatArrayFromBytes(vectorBytes)
vectors              = Array_.partition(vectors, 50)
console.log(`\tVectors: (${vectors.length}, ${vectors[0].length})`) 

// --- prepare indexer 
console.log("Preparing indexer.")
const indexer = new Annoy(500, 'Euclidean')
for(let i = 0; i < vectors.length; i++) {
    console.log(`---- Adding item ${i} of ${vectors.length}.`)
    const vector = vectors[i]
    indexer.addItem(i, normalize(vector))
}

// --- build indexer 
console.log("Building indexer.")
indexer.build()

// --- query indexer
console.log("Finding closest top 1, 10, 100, and 1000 words for all items.")
const top1    = []
const top10   = []
const top100  = []
const top1000 = []

for(let i = 0; i < vectors.length; i++) {
    console.log(`---- Finding closest items ${i + 1} of ${vectors.length}`)
    let results 

    results = indexer.getNNsByVector(vectors[i], 1001).slice(1)

    let closest = null
    let j = 0
    
    while(results[j] == i) {
        console.log("--- INCR")
        j += 1
    }

    closest = results[j]

    top1.push(closest)
    top10.push(results[9])
    top100.push(results[99])
    top1000.push(results[999])
}

// --- saving data files
console.log("Saving to files.")
let names = [1, 10, 100, 1000]
let i = 0
for(let items of [top1, top10, top100, top1000]) { 
    console.log("---- Processing for top " + names[i])
    const bytes = Encoder_.encodeIntArrayToBytes(items)
    fs.writeFileSync(OUTPUT_FOLDER + "/closest-" + names[i] + ".bin", bytes)
    i += 1
}

File: scripts/find_closest_words.py
from annoy import AnnoyIndex
import array 


VECTORS_FILE  = "./data/glove-50d/preprocessed/vectors.bin" 
INDEXER_FILE  = "./data/glove-50d/preprocessed/indexer.ann" 
OUTPUT_FOLDER = "./data/glove-50d/preprocessed/" 
DIM_COUNT    = 50
TREE_COUNT   = 100

# --- utility functions 
def chunks(lst, n):
    for i in range(0, len(lst), n):
        yield lst[i:i + n]

def encode_to_int_bytes(lst):
    items_b    = array.array("i", lst).tobytes() 
    return items_b


# --- load vectors file 
print("Loading vectors.")
vectors_ = open(VECTORS_FILE, "rb").read() 
vectors = array.array("f") 
vectors.frombytes(vectors_) 
vectors = vectors.tolist() 
vectors = list(chunks(vectors, DIM_COUNT))
print(f"\tShape: ({len(vectors)}, {len(vectors[0])})")

# --- load indexer 
print("Loading indexer.")
indexer = AnnoyIndex(DIM_COUNT, 'angular')
indexer.load(INDEXER_FILE)
print(f"---- Indexer: {indexer}")

# ---- find closest words 
print("Finding closest words.") 
closest = []
for i in range(len(vectors)): 
    print(f"---- Processing {i} of {len(vectors)}.")
    vector = vectors[i]
    results = indexer.get_nns_by_vector(vector, 10)
    j = 0 
    while results[j] == i:
        print("---- INCR")
        j += 1
    closest.append(results[j])

# --- saving to files 
print("Saving files.")

open(OUTPUT_FOLDER + "/closest.bin", "wb").write(encode_to_int_bytes(closest))

print("Done.")